### PR TITLE
Add ruby 3.2 and remove 2.4

### PR DESCRIPTION
**Pull request by KinWang-2013**
This PR adds ruby 3.2 and remove 2.4

### CODE REVIEW

File: .github/workflows/ruby.yml [hunk 1]
```diff
@@ -9,13 +9,11 @@ jobs:
       fail-fast: false
       matrix:
         include:
+          - ruby-version: 3.2
           - ruby-version: 3.1
           - ruby-version: 3.0
```
Comment (runlevel5) [inline .github/workflows/ruby.yml:14]: Why don't we remove 3.0 down to 2.4 since they have reached EOL?

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 * [#131](https://github.com/radar/distance_of_time_in_words/pull/131): Deprecates `highest_measure_only`. Adds alternate form of `highest_measures` option to permit rounding up whatever part of the duration was previously silently discarded - [@seansfkelley](https://github.com/seansfkelley).
 * [#133](https://github.com/radar/distance_of_time_in_words/pull/133): Test on Ruby 3.0 and 3.1 - [@dblock](https://github.com/dblock).
+* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Revmove ruby 2.4 and add ruby 3.2 - [@KinWang-2013](https://github.com/KinWang-2013)
```
Comment (runlevel5) [inline CHANGELOG.md:5]:
```suggestion
* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Remove ruby 2.4 and add ruby 3.2 - [@KinWang-2013](https://github.com/KinWang-2013)
```

Comment (KinWang-2013) [reply CHANGELOG.md:5]: thanks

File: .github/workflows/ruby.yml (hunk 1 above)
Comment (KinWang-2013) [reply .github/workflows/ruby.yml:14]: @dblock suggested we remove 2.4 for now

Comment (dblock) [reply .github/workflows/ruby.yml:14]: I am not opposed to removing everything that's EOL.

File: .github/workflows/ruby.yml [hunk 3]
```diff
@@ -9,13 +9,11 @@ jobs:
...
           - ruby-version: 2.5
           - ruby-version: 2.6
-          - ruby-version: 2.4
-            bundler-version: 1.17.3
-            gemfile: gemfiles/rails_4.gemfile
```
Comment (dblock) [inline .github/workflows/ruby.yml:18]: If we're no longer testing Rails 4, delete this Gemfile.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 * [#131](https://github.com/radar/distance_of_time_in_words/pull/131): Deprecates `highest_measure_only`. Adds alternate form of `highest_measures` option to permit rounding up whatever part of the duration was previously silently discarded - [@seansfkelley](https://github.com/seansfkelley).
 * [#133](https://github.com/radar/distance_of_time_in_words/pull/133): Test on Ruby 3.0 and 3.1 - [@dblock](https://github.com/dblock).
+* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Remove ruby 2.4 and add ruby 3.2 - [@KinWang-2013](https://github.com/KinWang-2013)
```
Comment (dblock) [inline CHANGELOG.md:5]:
```suggestion
* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Add support for Ruby 3.2 - [@KinWang-2013](https://github.com/KinWang-2013).
* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Remove support for Ruby 2.4 and Rails 4 - [@KinWang-2013](https://github.com/KinWang-2013).
```

**Review by dblock — CHANGES_REQUESTED** [review]
Let's do this.

**Comment by KinWang-2013** [comment]
@dblock @runlevel5 please check again, made the changes

**Review by dblock — CHANGES_REQUESTED** [review]
I don't think we should be removing tests for all versions of Rails and not for Ruby 3. Can we bring back the Rails Gemfiles on a newer version of Ruby? Also see if we can get 2.7 to work.

**Comment by KinWang-2013** [comment]
> I don't think we should be removing tests for all versions of Rails and not for Ruby 3. Can we bring back the Rails Gemfiles on a newer version of Ruby? Also see if we can get 2.7 to work.

 you mean we keep ruby 2.7 onwards ?

**Comment by KinWang-2013** [comment]
> > you mean we keep ruby 2.7 onwards ?
> 
> Yes, just to have support for 2.x, unless it's difficult.

okay I will try once

**Comment by KinWang-2013** [comment]
@dblock I made the updates, currently with ruby 3.2, rails 5.x tests are failing, narrowed down to 

```
# file: lib/dotiw/      
141		I18n.with_options locale: options[:locale], scope: i18n_scope do |locale|
 142			phrases = hash.map { |key, value| locale.t(key, count: value) }
 143  	end
```
this block which calls I18n.t method

I am still looking into the issue, if you have any advice, it would be helpful

**Comment by dblock** [comment]
> @dblock I made the updates, currently with ruby 3.2, rails 5.x tests are failing, narrowed down to
> 
> ```
> # file: lib/dotiw/methods.rb
>  141		I18n.with_options locale: options[:locale], scope: i18n_scope do |locale|
>  142			phrases = hash.map { |key, value| locale.t(key, count: value) }
>  143  	end
> ```
> 
> this block which calls I18n.t method
> 
> I am still looking into the issue, if you have any advice, it would be helpful

I spent some time on this and I am as puzzled as you. It's some combination of Ruby 3.x keyword arguments with the way this method is declared that's not adding up.

It works with Ruby 2.7, so keep Rails 5 with just Ruby 2.7 tests and we should be good.

**Comment by KinWang-2013** [comment]
> > @dblock I made the updates, currently with ruby 3.2, rails 5.x tests are failing, narrowed down to
> > ```
> > # file: lib/dotiw/methods.rb
> >  141		I18n.with_options locale: options[:locale], scope: i18n_scope do |locale|
> >  142			phrases = hash.map { |key, value| locale.t(key, count: value) }
> >  143  	end
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > this block which calls I18n.t method
> > I am still looking into the issue, if you have any advice, it would be helpful
> 
> I spent some time on this and I am as puzzled as you. It's some combination of Ruby 3.x keyword arguments with the way this method is declared that's not adding up.
> 
> It works with Ruby 2.7, so keep Rails 5 with just Ruby 2.7 tests and we should be good.

okay cool I can do that.

BTW I also spent some time to debug the issue and found out that it is related to `with_options` method in the file `active_support/core_ext/object/with_options.rb` from ActiveSupport. Upon further inspection, it seems like when we use with options, it passes the options as a hash parameter when we call `locale.t` hence the `ArgumentError` from I18n gem. I tried to remove the with_options and call the method directly but some tests are failing (ActionView tests) because the ActionView date helper uses similar code in line 104 of file `action_view/helpers/date_helper.rb` (V5.0.7.2), so I did left it as is.

**Comment by KinWang-2013** [comment]
@dblock made the suggested changes, all builds are passing as of now, please check

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,12 @@
...
 * [#133](https://github.com/radar/distance_of_time_in_words/pull/133): Test on Ruby 3.0 and 3.1 - [@dblock](https://github.com/dblock).
+* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Add support for Ruby 3.2 - [@KinWang-2013](https://github.com/KinWang-2013).
+* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Remove support for Ruby 2.4, 2.5 and 2.6 - [@KinWang-2013](https://github.com/KinWang-2013).
+* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Add support for Ruby 2.7 - [@KinWang-2013](https://github.com/KinWang-2013).
+* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Add test for Rails 6.0 with Ruby 3.2 - [@KinWang-2013](https://github.com/KinWang-2013).
+* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Add test for Rails 5.X with Ruby 2.7 - [@KinWang-2013](https://github.com/KinWang-2013).
+* [#135](https://github.com/radar/distance_of_time_in_words/pull/135): Remove support for Rails 4 - [@KinWang-2013](https://github.com/KinWang-2013).
```
Comment (dblock) [inline CHANGELOG.md:10]: Can we collapse these?

```
- Add support for Ruby 2.7 and 3.2, removed support for ruby 2.4, 2.5, 2.6 and Rails 4
```

No need to mention that we added tests.

**Comment by dblock** [comment]
> @dblock made the suggested changes, all builds are passing as of now, please check

Looks good, just one more annoying ask from me on CHANGELOG, thanks for hanging in here with me!

**Comment by KinWang-2013** [comment]
> > @dblock made the suggested changes, all builds are passing as of now, please check
> 
> Looks good, just one more annoying ask from me on CHANGELOG, thanks for hanging in here with me!

no worries, made the changes, please check

**Comment by dblock** [comment]
Good work, merged. Add Rails 7 in another PR?

**Comment by KinWang-2013** [comment]
> Good work, merged. Add Rails 7 in another PR?

okay

**Pull request merged**